### PR TITLE
Update measurement link examples to match dcpower example changes.

### PR DESCRIPTION
**Pull request by subash-suresh**
- [x] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/measurement-services-python/blob/main/CONTRIBUTING.md).

TODO: Check the above box with an 'x' indicating you've read and followed [CONTRIBUTING.md](https://github.com/ni/measurement-services-python/blob/main/CONTRIBUTING.md).

### What does this Pull Request accomplish?

This Pull request updates the measurement link examples to match the changes recently done in the NI DC power examples.
The updates are as follows,
1. Created new driver specific helper files which contains the session creation logic.
2. Update the general helper files to match the updates done in ni dcpower example.
3. Update the reservation and session creation logic to match the updates done in ni dcpower example.

### Why should this Pull Request be merged?

Updates the measurement link example files to match the latest changes done to measurement service and the helpers.

### What testing has been done?

Manually ran all the tests and verified the functionality of them.

### CODE REVIEW

**Comment by subash-suresh** [comment]
I would like to get some thoughts on moving the helpers.py into the measurement service module, so that it could be accessible by all the examples and measurements using them.

**Comment by dixonjoel** [comment]
> I would like to get some thoughts on moving the helpers.py into the measurement service module, so that it could be accessible by all the examples and measurements using them.

The argument that I've heard against doing that is we would have to maintain that code with backwards compatibility going forward and that we didn't want to take that responsibility on for example code.
@bkeryan may have some other thoughts.

**Comment by bkeryan** [comment]
> The argument that I've heard against doing that is we would have to maintain that code with backwards compatibility going forward and that we didn't want to take that responsibility on for example code. @bkeryan may have some other thoughts.

That's exactly right.

File: examples/nidaqmx_analog_input/measurement.py [hunk 1]
```diff
@@ -46,12 +49,7 @@ def measure(pin_name, sample_rate, number_of_samples):
         sample_rate,
```
Comment (bkeryan) [inline examples/nidaqmx_analog_input/measurement.py:49]: def measure is missing type annotations.
```
measurement.py:71: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
```

**Review by bkeryan — CHANGES_REQUESTED** [review]
Many of these issues apply to multiple examples.

File: examples/nidaqmx_analog_input/_nidaqmx_helpers.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
+"""nidaqmx Helper classes and functions for MeasurementLink examples."""
+
+import grpc
+import nidaqmx
+
+import ni_measurementlink_service as nims
+
+
+def _create_nidaqmx_task(
```
[the remark this answers is not in the document]
Comment (subash-suresh) [reply examples/nidaqmx_analog_input/_nidaqmx_helpers.py]: updated as per suggestions

File: examples/nidigital_spi/_nidigital_helpers.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+from typing import Any, Dict
+
+import grpc
+import nidigital
+
+import ni_measurementlink_service as nims
+
+# To use a physical NI-DIGITAL instrument, set this to False or specify
+# --no-use-simulation on the command line.
+USE_SIMULATION = True
```
[the remark this answers is not in the document]
Comment (subash-suresh) [reply examples/nidigital_spi/_nidigital_helpers.py]: udpated as per suggestions

File: examples/nidmm_measurement/_nidmm_helpers.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+from typing import Any, Dict
+
+import grpc
+import nidmm
+
+import ni_measurementlink_service as nims
+
+# To use a physical NI-DMM instrument, set this to False or specify
+# --no-use-simulation on the command line.
+USE_SIMULATION = True
```
[the remark this answers is not in the document]
Comment (subash-suresh) [reply examples/nidmm_measurement/_nidmm_helpers.py]: Updated as per suggestions

File: examples/nidigital_spi/measurement.py [hunk 5]
```diff
@@ -3,19 +3,22 @@
...
+    get_grpc_device_channel,
     get_service_options,
     grpc_device_options,
-    verbosity_option,
     use_simulation_option,
+    verbosity_option,
 )
+from _nidigital_helpers import _create_nidigital_session
 
 import ni_measurementlink_service as nims
```
[the remark this answers is not in the document]
Comment (subash-suresh) [reply examples/nidigital_spi/measurement.py:24]: Removed the duplicated USE_SIMULATION definition.

File: examples/nidaqmx_analog_input/measurement.py [hunk 6]
```diff
@@ -46,12 +49,7 @@ def measure(pin_name, sample_rate, number_of_samples):
...
         number_of_samples,
     )
-    session_management_client = nims.session_management.Client(
-        grpc_channel=measurement_service.get_channel(
-            provided_interface=nims.session_management.GRPC_SERVICE_INTERFACE_NAME,
-            service_class=nims.session_management.GRPC_SERVICE_CLASS,
-        )
-    )
+    session_management_client = create_session_management_client(measurement_service)
     with contextlib.ExitStack() as stack:
```
[the remark this answers is not in the document]
Comment (subash-suresh) [reply examples/nidaqmx_analog_input/measurement.py]: updated all the single session examples as per suggestions

File: examples/nidaqmx_analog_input/_nidaqmx_helpers.py [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+    session_info: nims.session_management.SessionInformation,
+    session_grpc_channel: grpc.Channel = None,
```
[the remark this answers is not in the document]
Comment (subash-suresh) [reply examples/nidaqmx_analog_input/_nidaqmx_helpers.py]: I have made the grpc channel parameter as optional through type hinting to solve this issue

File: examples/nidaqmx_analog_input/measurement.py [hunk 8]
```diff
@@ -80,7 +78,8 @@ def cancel_callback():
 
         measurement_service.context.add_cancel_callback(cancel_callback)
 
-        task = stack.enter_context(_create_nidaqmx_task(session_info))
+        grpc_device_channel = get_grpc_device_channel(measurement_service, nidaqmx)
```
[the remark this answers is not in the document]
Comment (subash-suresh) [reply examples/nidaqmx_analog_input/measurement.py]: I have updated the `get_grpc_device_channel` to work based of the service options.

File: examples/nidaqmx_analog_input/_nidaqmx_helpers.py
[the remark this answers is not in the document]
Comment (subash-suresh) [reply examples/nidaqmx_analog_input/_nidaqmx_helpers.py:1]: Updated as per suggestions

File: examples/nidaqmx_analog_input/measurement.py (hunk 1 above)
Comment (subash-suresh) [reply examples/nidaqmx_analog_input/measurement.py:48]: I will take this in a new PR.

File: examples/nidaqmx_analog_input/_nidaqmx_helpers.py
Comment (bkeryan) [inline examples/nidaqmx_analog_input/_nidaqmx_helpers.py:1]: Updated the teststand fixtures to use the new helpers. This applies to every example except `nidcpower_source_dc_voltage` which has already been updated.

File: examples/nidaqmx_analog_input/_helpers.py [hunk 9]
```diff
@@ -231,3 +236,30 @@ def use_simulation_option(default: bool) -> Callable[[F], F]:
...
+    driver_module: types.ModuleType,
+    service_options: ServiceOptions,
+) -> grpc.Channel:
+    """Returns driver specific grpc device channel."""
+    if service_options.use_grpc_device and service_options.grpc_device_address:
+        return measurement_service.get_channel(service_options.grpc_device_address)
+
+    return measurement_service.get_channel(
+        provided_interface=getattr(driver_module, "GRPC_SERVICE_INTERFACE_NAME"),
+        service_class="ni.measurementlink.v1.grpcdeviceserver",
```
Comment (bkeryan) [inline examples/nidaqmx_analog_input/_helpers.py:252]: This should return None when service_options.use_grpc_device is false.

Update the return type annotation too.

File: examples/nidigital_spi/_nidigital_helpers.py [hunk 10]
```diff
@@ -0,0 +1,34 @@
+"""nidigital Helper classes and functions for MeasurementLink examples."""
+
+from typing import Any, Dict, Optional
+
+import grpc
+import nidigital
+
+import ni_measurementlink_service as nims
+
+# To use a physical NI Digital Pattern instrument, set this to False or specify
```
Comment (bkeryan) [inline examples/nidigital_spi/_nidigital_helpers.py:10]:
```suggestion
# To use a physical NI digital pattern instrument, set this to False or specify
```

File: examples/nivisa_dmm_measurement/measurement.py [hunk 11]
```diff
@@ -77,12 +78,7 @@ def measure(
...
 
-    session_management_client = nims.session_management.Client(
-        grpc_channel=measurement_service.get_channel(
-            provided_interface=nims.session_management.GRPC_SERVICE_INTERFACE_NAME,
-            service_class=nims.session_management.GRPC_SERVICE_CLASS,
-        )
-    )
+    session_management_client = create_session_management_client(measurement_service)
 
     with contextlib.ExitStack() as stack:
```
Comment (bkeryan) [inline examples/nivisa_dmm_measurement/measurement.py:83]: Remove exit stack

**Review by bkeryan — CHANGES_REQUESTED** [review]
Oops, we overlooked the TestStand fixtures, which are the reason we're adding the create_session helpers in the first place.

File: examples/nivisa_dmm_measurement/measurement.py (hunk 11 above)
Comment (subash-suresh) [reply examples/nivisa_dmm_measurement/measurement.py]: removed exit stack

File: examples/nidaqmx_analog_input/_helpers.py (hunk 9 above)
Comment (subash-suresh) [reply examples/nidaqmx_analog_input/_helpers.py]: Updated as per suggestions and included type annotations for the return type

File: examples/nidaqmx_analog_input/_nidaqmx_helpers.py
Comment (subash-suresh) [reply examples/nidaqmx_analog_input/_nidaqmx_helpers.py:1]: Yes, I have updated all the teststand fixtures to use the latest updates and tested them as well

File: examples/nivisa_dmm_measurement/measurement.py [hunk 12]
```diff
@@ -101,29 +96,27 @@ def measure(
...
+        with create_visa_session(resource_manager, session_info.resource_name) as session:
+            # Work around https://github.com/pyvisa/pyvisa/issues/739 - Type annotation for Resource
+            # context manager implicitly upcasts derived class to base class
+            assert isinstance(session, pyvisa.resources.MessageBasedResource)
+
+            log_instrument_id(session)
+
+            # When this measurement is called from outside of TestStand (session_exists == False),
+            # reset the instrument to a known state.
+            # In TestStand, ProcessSetup resets the instrument.
```
Comment (bkeryan) [inline examples/nivisa_dmm_measurement/measurement.py:108]: Why did you split this line? It's not >100 columns.

File: examples/nidaqmx_analog_input/teststand_fixture.py [hunk 13]
```diff
@@ -60,14 +61,14 @@ def create_nidaqmx_tasks(sequence_context: Any) -> None:
             timeout=0,
         ) as reservation:
             for session_info in reservation.session_info:
-                session_kwargs["grpc_options"] = GrpcSessionOptions(
-                    grpc_channel_pool.get_grpc_device_channel(GRPC_SERVICE_INTERFACE_NAME),
-                    session_name=session_info.session_name,
-                    initialization_behavior=SessionInitializationBehavior.INITIALIZE_SERVER_SESSION,
+                grpc_device_channel = grpc_channel_pool.get_grpc_device_channel(
+                    nidaqmx.GRPC_SERVICE_INTERFACE_NAME
+                )
```
Comment (bkeryan) [inline examples/nidaqmx_analog_input/teststand_fixture.py:59]: Overlooked in NI-DCPower PR review: get_grpc_device_channel doesn't depend on session_info, so hoist it out of the loop.

File: examples/nidmm_measurement/teststand_fixture.py [hunk 14]
```diff
@@ -55,22 +56,13 @@ def create_nidmm_sessions(sequence_context: Any) -> None:
...
                 )
-
-                # Leave session open
-                nidmm.Session(
-                    resource_name=session_info.resource_name,
-                    options=options,
-                    grpc_options=grpc_options,
+                create_session(
+                    session_info,
+                    grpc_device_channel,
```
Comment (bkeryan) [inline examples/nidmm_measurement/teststand_fixture.py:64]: Overlooked in NI-DCPower PR review: where did the comment go? Also, a discard would make it more explicit that we are ignoring the return value.

```
# Leave session open
_ = create_session(
```

**Review by bkeryan — CHANGES_REQUESTED** [review]

File: examples/nivisa_dmm_measurement/measurement.py (hunk 12 above)
Comment (subash-suresh) [reply examples/nivisa_dmm_measurement/measurement.py:108]: There are 102 columns and I encountered a lint error hence I moved it to the next line.

File: examples/nidaqmx_analog_input/teststand_fixture.py (hunk 13 above)
Comment (subash-suresh) [reply examples/nidaqmx_analog_input/teststand_fixture.py:59]: I have moved it out of the loop, and did similar changes to other examples as well.

File: examples/nidaqmx_analog_input/_helpers.py [hunk 15]
```diff
@@ -242,15 +242,17 @@ def get_grpc_device_channel(
...
     driver_module: types.ModuleType,
     service_options: ServiceOptions,
-) -> grpc.Channel:
+) -> Union[grpc.Channel, None]:
     """Returns driver specific grpc device channel."""
-    if service_options.use_grpc_device and service_options.grpc_device_address:
-        return measurement_service.get_channel(service_options.grpc_device_address)
+    if service_options.use_grpc_device:
+        if service_options.grpc_device_address:
+            return measurement_service.get_channel(service_options.grpc_device_address)
```
[the remark this answers is not in the document]
Comment (subash-suresh) [reply examples/nidaqmx_analog_input/_helpers.py]: updated the helpers as per suggestions

File: examples/nidaqmx_analog_input/_helpers.py [hunk 16]
```diff
@@ -242,15 +242,17 @@ def get_grpc_device_channel(
     measurement_service: MeasurementService,
     driver_module: types.ModuleType,
     service_options: ServiceOptions,
-) -> grpc.Channel:
+) -> Union[grpc.Channel, None]:
```
[the remark this answers is not in the document]
Comment (subash-suresh) [reply examples/nidaqmx_analog_input/_helpers.py]: I have updated to use `Optional`

File: examples/nivisa_dmm_measurement/measurement.py (hunk 12 above)
Comment (bkeryan) [reply examples/nivisa_dmm_measurement/measurement.py:108]: Oops, I was looking at the wrapped version. 

Wrapping the last word is sufficient. The VS Code "Rewrap" extension is good for this.

**Review by bkeryan — APPROVED** [review]
Approved with suggestions

File: examples/nidaqmx_analog_input/teststand_fixture.py [hunk 17]
```diff
@@ -60,14 +54,14 @@ def create_nidaqmx_tasks(sequence_context: Any) -> None:
...
```
[the remark this answers is not in the document]
Comment (subash-suresh) [reply examples/nidaqmx_analog_input/teststand_fixture.py]: moved it outside the loop

File: examples/nidcpower_source_dc_voltage/teststand_fixture.py [hunk 18]
```diff
@@ -46,24 +46,25 @@ def create_nidcpower_sessions(sequence_context: Any) -> None:
...
-                )
-                create_session(
+                # Leave session open
+                _ = create_session(
                     session_info,
                     grpc_device_channel,
                     initialization_behavior=nidcpower.SessionInitializationBehavior.INITIALIZE_SERVER_SESSION,
                 )
-
             session_management_client.register_sessions(reservation.session_info)
```
[the remark this answers is not in the document]
Comment (subash-suresh) [reply examples/nidcpower_source_dc_voltage/teststand_fixture.py:70]: moved outside the loop

File: examples/nidaqmx_analog_input/teststand_fixture.py [hunk 19]
```diff
@@ -88,13 +82,12 @@ def destroy_nidaqmx_tasks() -> None:
             session_management_client.unregister_sessions(reservation.session_info)
 
             for session_info in reservation.session_info:
-                grpc_options = GrpcSessionOptions(
-                    grpc_channel_pool.get_grpc_device_channel(GRPC_SERVICE_INTERFACE_NAME),
-                    session_name=session_info.session_name,
-                    initialization_behavior=SessionInitializationBehavior.ATTACH_TO_SERVER_SESSION,
+                grpc_device_channel = grpc_channel_pool.get_grpc_device_channel(
+                    nidaqmx.GRPC_SERVICE_INTERFACE_NAME
                 )
```
[the remark this answers is not in the document]
Comment (subash-suresh) [reply examples/nidaqmx_analog_input/teststand_fixture.py]: moved outside the loop

**Pull request merged**